src/unitycatalog/ai/utils.py

```diff
@@ -113,3 +141,148 @@ def validate_full_function_name(function_name: str) -> FullFunctionName:
+        elif tpe == "map":
+            key_type = uc_type_json["keyType"]
+            if key_type != "string":
+                raise TypeError(f"Only support STRING key type for MAP but got {key_type}.")
+            value_type = uc_type_json_to_pydantic_type(uc_type_json["valueType"])
+            if uc_type_json["valueContainsNull"]:
+                value_type = Optional[value_type]
+            return Dict[str, value_type]
+        elif tpe == "struct":
+            fields = {}
+            for field in uc_type_json["fields"]:
+                field_type = uc_type_json_to_pydantic_type(field["type"])
+                comment = field.get("metadata", {}).get("comment")
+                if field.get("nullable"):
+                    field_type = Optional[field_type]
+                    fields[field["name"]] = (field_type, Field(default=None, description=comment))
+                else:
+                    fields[field["name"]] = (field_type, Field(..., description=comment))
+            uc_type_json_str = json.dumps(uc_type_json, sort_keys=True)
+            type_hash = md5(uc_type_json_str.encode()).hexdigest()[:8]
+            return create_model(f"Struct_{type_hash}", **fields)
+    raise TypeError(f"Unknown type {uc_type_json}.")
+
+
+# TODO: add UC OSS support
+def supported_param_info_types():
+    types = ()
+    try:
+        from databricks.sdk.service.catalog import FunctionParameterInfo
+
+        types += (FunctionParameterInfo,)
+    except ImportError:
+        pass
+
+    return types
+
+
+# TODO: add UC OSS support
+def supported_function_info_types():
+    types = ()
+    try:
+        from databricks.sdk.service.catalog import FunctionInfo
+
+        types += (FunctionInfo,)
+    except ImportError:
+        pass
+
+    return types
+
+
+@dataclass
+class PydanticField:
+    pydantic_type: Type
+    description: Optional[str] = None
+    default: Optional[Any] = None
+
+
+def param_info_to_pydantic_type(param_info: Any) -> PydanticField:
+    """
+    Convert Unity Catalog function parameter information to Pydantic type.
+
+    Args:
+        param_info: The Unity Catalog function parameter information.
+            It must be either databricks.sdk.service.catalog.FunctionParameterInfo or
+            unitycatalog.types.function_info.InputParamsParameter object.
+    """
+    if not isinstance(param_info, supported_param_info_types()):
+        raise TypeError(f"Unsupported parameter info type: {type(param_info)}")
+    if param_info.type_json is None:
+        raise ValueError(f"Parameter type json is None for parameter {param_info.name}.")
+    type_json = json.loads(param_info.type_json)
+    nullable = type_json.get("nullable")
+    pydantic_type = uc_type_json_to_pydantic_type(type_json["type"])
+    if nullable:
+        pydantic_type = Optional[pydantic_type]
+    default = None
+    description = param_info.comment or ""
+    if param_info.parameter_default:
+        pydantic_type = Optional[pydantic_type]
```

Comment:
Do we need `Optional`?

```python
class MyModel:
    s: str = "foo"
```

`Optional` should not be used if the default value is present.